moshi-kotlin/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapterFactory.kt

```diff
@@ -213,111 +225,102 @@ public class KotlinJsonAdapterFactory : JsonAdapter.Factory {
       // Fall back to a reflective adapter when the generated adapter is not found.
     }
 
-    require(!rawType.isLocalClass) {
-      "Cannot serialize local class or object expression ${rawType.name}"
-    }
-    val rawTypeKotlin = rawType.kotlin
-    require(!rawTypeKotlin.isAbstract) {
+    require(!rawType.isLocalClass) { "Cannot serialize local class or object expression ${rawType.name}" }
+
+    require(!rawType.isAnonymousClass) { "Cannot serialize anonymous class ${rawType.name}" }
+
+    val kmClass = rawType.header()?.toKmClass() ?: return null
+
+    require(kmClass.modality != Modality.ABSTRACT) {
       "Cannot serialize abstract class ${rawType.name}"
     }
-    require(!rawTypeKotlin.isInner) {
-      "Cannot serialize inner class ${rawType.name}"
-    }
-    require(rawTypeKotlin.objectInstance == null) {
+    require(!kmClass.isInner) { "Cannot serialize inner class ${rawType.name}" }
+    require(kmClass.kind != ClassKind.OBJECT) {
       "Cannot serialize object declaration ${rawType.name}"
     }
-    require(!rawTypeKotlin.isSealed) {
+    require(kmClass.kind != ClassKind.COMPANION_OBJECT) {
+      "Cannot serialize companion object declaration ${rawType.name}"
+    }
+    require(kmClass.modality != Modality.SEALED) {
       "Cannot reflectively serialize sealed class ${rawType.name}. Please register an adapter."
     }
 
-    val constructor = rawTypeKotlin.primaryConstructor ?: return null
-    val parametersByName = constructor.parameters.associateBy { it.name }
-    constructor.isAccessible = true
+    val ktConstructor = KtConstructor.primary(rawType, kmClass) ?: return null
 
-    val bindingsByName = LinkedHashMap<String, KotlinJsonAdapter.Binding<Any, Any?>>()
+    // TODO CR this doesn't cover platform types
```

Comment:
I can add this, I just want to check if we really want to support this as code gen does not support this (explicitly) and there is a test against it. Commented on an example test of this in this PR too